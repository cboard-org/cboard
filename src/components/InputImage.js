import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import RaisedButton from 'material-ui/RaisedButton';
import FontIcon from 'material-ui/FontIcon';

const styles = {
  imageInput: {
    cursor: 'pointer',
    position: 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
    left: 0,
    width: '100%',
    opacity: 0,
  },
};

const getBase64Image = (img, width = img.width, height = img.height) => {
  const canvas = document.createElement('canvas');
  canvas.width = width;
  canvas.height = height;

  const ctx = canvas.getContext('2d');
  ctx.drawImage(img, 0, 0, width, height);

  const dataURL = canvas.toDataURL('image/png');
  return dataURL;
};

class InputImage extends PureComponent {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const file = event.target.files[0];
    const reader = new FileReader();

    reader.onload = (ev) => {
      const img = document.createElement('img');
      const width = 512;
      const height = 512;

      img.onload = (e) => {
        const imageData = getBase64Image(e.target, width, height);
        this.props.onChange(imageData);
      };
      img.src = ev.target.result;
    };
    reader.readAsDataURL(file);
    this.input.value = '';
  }

  render() {
    return (
      <div>
        <RaisedButton
          label="Choose an Image"
          containerElement="label"
          icon={<FontIcon className="material-icons">add_a_photo</FontIcon>}
        >
          <input type="file" style={styles.imageInput} onChange={this.handleChange} ref={(input) => { this.input = input; }} />
        </RaisedButton>
      </div >
    );
  }
}

export default InputImage;

InputImage.propTypes = {
  onChange: PropTypes.func.isRequired,
};
