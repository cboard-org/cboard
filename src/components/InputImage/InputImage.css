.InputImage {
  position: relative;
  display: block;
  width: 144px;
  height: 144px;
  background: #ddd;
  cursor: pointer;
  margin: 16px auto 0;
}

.InputImage > svg {
  width: 32px;
  height: 32px;
  padding: 4px;
  margin: 12px;
  border-radius: 50%;
  position: absolute;
  bottom: 0;
  right: 0;
  background: #fff;
}

.InputImage__label {
  display: block;
  text-align: center;
  line-height: 120px;
  height: 100%;
  cursor: pointer;
}

.InputImage__label.is-uploaded {
  height: 0;
}

.InputImage__input {
  display: none;
}

.InputImage__img {
  height: 100%;
  width: 100%;
}
