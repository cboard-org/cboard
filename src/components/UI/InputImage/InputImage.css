.InputImage {
  position: relative;
  display: flex;
  align-items: center;
  width: 96px;
  height: 96px;
  background: #ddd;
  cursor: pointer;
}

@media (min-width: 600px) {
  .InputImage {
    width: 144px;
    height: 144px;
  }
}

.InputImage > svg {
  width: 32px;
  height: 32px;
  padding: 4px;
  margin: 12px;
  border-radius: 50%;
  position: absolute;
  bottom: 0;
  right: 0;
  color: #222;
  background: #fff;
  pointer-events: none;
}

.InputImage__label {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  text-align: center;
  cursor: pointer;
}

.InputImage.is-uploaded .InputImage__label {
  opacity: 0;
}

.InputImage__input {
  display: none;
}

.InputImage__img {
  object-fit: contain;
  width: 100%;
  height: 100%;
}
