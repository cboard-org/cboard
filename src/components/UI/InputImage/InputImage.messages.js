import { defineMessages } from 'react-intl';

export default defineMessages({
  uploadImage: {
    id: 'cboard.components.InputImage.uploadImage',
    defaultMessage: 'Upload image'
  }
});
