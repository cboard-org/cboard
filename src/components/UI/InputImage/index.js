export { default } from './InputImage.component';
