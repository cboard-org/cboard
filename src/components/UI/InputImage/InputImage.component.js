import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import PhotoCameraIcon from '@material-ui/icons/PhotoCamera';
import readAndCompressImage from 'browser-image-resizer';
import CircularProgress from '@material-ui/core/CircularProgress';

import API from '../../../api';
import messages from './InputImage.messages';
import './InputImage.css';

class InputImage extends PureComponent {
  static propTypes = {
    /**
     * @ignore
     */
    intl: intlShape.isRequired,
    user: PropTypes.object,
    /**
     * Callback fired when input changes
     */
    onChange: PropTypes.func.isRequired
  };

  state = {
    loading: false
  };

  blobToBase64(blob) {
    const reader = new FileReader();

    reader.addEventListener(
      'load',
      () => {
        this.props.onChange(reader.result);
      },
      false
    );

    reader.readAsDataURL(blob);
  }

  async resizeImage(
    file,
    config = {
      quality: 7,
      maxWidth: 200,
      maxHeight: 200,
      autoRotate: true,
      debug: false
    }
  ) {
    const resizedImage = await readAndCompressImage(file, config);
    return resizedImage;
  }

  handleChange = async event => {
    const { onChange, user } = this.props;
    const file = event.target.files[0];
    const resizedImage = await this.resizeImage(file);
    // Loggedin user?
    if (user) {
      this.setState({
        loading: true
      });
      try {
        const imageUrl = await API.uploadFile(resizedImage, file.name);
        onChange(imageUrl);
      } catch (error) {
        const imageBase64 = this.blobToBase64(resizedImage);
        onChange(imageBase64);
      } finally {
        this.setState({
          loading: false
        });
      }
    } else {
      const imageBase64 = this.blobToBase64(resizedImage);
      onChange(imageBase64);
    }
  };

  render() {
    const { intl } = this.props;

    return (
      <div className="InputImage">
        <PhotoCameraIcon />
        {this.state.loading ?
          <CircularProgress
            size={24}
            thickness={7}
          />
          : <label className="InputImage__label">
            {intl.formatMessage(messages.uploadImage)}
            <input
              className="InputImage__input"
              type="file"
              accept="image/*"
              onChange={this.handleChange}
            />
          </label>
        }
      </div>
    );
  }
}

export default injectIntl(InputImage);
