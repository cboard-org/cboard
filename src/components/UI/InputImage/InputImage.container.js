import React from 'react';
import { connect } from 'react-redux';
import InputImage from './InputImage.component';

const InputImageContainer = props => <InputImage {...props} />;

const mapStateToProps = ({ app: { userData } }, ownProps) => ({
  ...ownProps,
  user: userData.email ? userData : null
});

export default connect(mapStateToProps)(InputImageContainer);
