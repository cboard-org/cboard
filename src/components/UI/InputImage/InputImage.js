import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import PhotoCameraIcon from '@material-ui/icons/PhotoCamera';
import readAndCompressImage from 'browser-image-resizer';

import messages from './InputImage.messages';
import './InputImage.css';

class InputImage extends PureComponent {
  static propTypes = {
    /**
     * @ignore
     */
    intl: intlShape.isRequired,
    /**
     * Callback fired when input changes
     */
    onChange: PropTypes.func.isRequired
  };

  blobToBase64(blob) {
    const reader = new FileReader();

    reader.addEventListener(
      'load',
      () => {
        this.props.onChange(reader.result);
      },
      false
    );

    reader.readAsDataURL(blob);
  }

  async resizeImage(
    file,
    config = {
      quality: 7,
      maxWidth: 200,
      maxHeight: 200,
      autoRotate: true,
      debug: false
    }
  ) {
    const resizedImage = await readAndCompressImage(file, config);
    return resizedImage;
  }

  handleChange = async event => {
    const { onChange } = this.props;
    const file = event.target.files[0];
    const resizedImage = await this.resizeImage(file);
    const imageBase64 = this.blobToBase64(resizedImage);
    onChange(imageBase64);
  };

  render() {
    const { intl } = this.props;

    return (
      <div className="InputImage">
        <label className="InputImage__label">
          {intl.formatMessage(messages.uploadImage)}
          <input
            className="InputImage__input"
            type="file"
            accept="image/*"
            onChange={this.handleChange}
          />
        </label>
        <PhotoCameraIcon />
      </div>
    );
  }
}

export default injectIntl(InputImage);
