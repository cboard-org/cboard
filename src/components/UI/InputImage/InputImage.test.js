import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import { mount, shallow } from 'enzyme';
import InputImage from './InputImage.component';

jest.mock('./InputImage.messages', () => {
  return {
    uploadImage: {
      id: 'cboard.components.InputImage.uploadImage',
      defaultMessage: 'Upload image'
    }
  };
});

describe('InputImage tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<InputImage onChange={() => {}} />);
  });

  test('on buttton click', () => {
    const wrapper = mount(
      shallow(<InputImage disabled={false} onChange={() => {}} />).get(0)
    );
    wrapper.simulate('click');
    expect(wrapper.state().loading).toEqual(false);
  });
});
