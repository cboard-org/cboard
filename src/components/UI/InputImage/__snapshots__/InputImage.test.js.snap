// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputImage tests default renderer 1`] = `
<InputImage
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
  onChange={[Function]}
/>
`;
